error: "Resource" has no attribute "destroy"

This error is caused by the following code in cli.py:

```python
def cli(...):
    if release is True:
        resources_class = iocage.lib.Releases.ReleasesGenerator
    else:
        resources_class = iocage.lib.Jails.JailsGenerator

...

    for item in resources:
    ....
        item.destroy()
```

`ReleasesGenerator` and `JailsGenerator` both inherit from (eventually) `Recourse` but `Resource` has no method `destroy()`.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/49958052-error-resource-has-no-attribute-destroy?utm_campaign=plugin&utm_content=tracker%2F70027116&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F70027116&utm_medium=issues&utm_source=github).
</bountysource-plugin>

```
iocage/lib/Host.py:98:28: error: Cannot instantiate abstract class 'DefaultResource' with abstract attribute 'destroy'
iocage/cli/stop.py:54:16: error: Cannot instantiate abstract class 'JailsGenerator' with abstract attribute 'destroy'
iocage/cli/start.py:68:16: error: Cannot instantiate abstract class 'JailsGenerator' with abstract attribute 'destroy'
iocage/cli/start.py:84:12: error: Cannot instantiate abstract class 'JailsGenerator' with abstract attribute 'destroy'
iocage/cli/set.py:66:16: error: Cannot instantiate abstract class 'JailsGenerator' with abstract attribute 'destroy'
iocage/cli/list.py:106:20: error: Cannot instantiate abstract class 'ReleasesGenerator' with abstract attribute 'destroy'
iocage/cli/destroy.py:63:8: error: Incompatible types in assignment (expression has type Type[JailsGenerator], variable has type Type[ReleasesGenerator])
iocage/cli/destroy.py:65:16: error: Cannot instantiate abstract class 'ReleasesGenerator' with abstract attribute 'destroy'
iocage/cli/destroy.py:93:8: error: Call to untyped function "destroy" of "Resource" in typed context
```

--- a/iocage/lib/Resource.py
+++ b/iocage/lib/Resource.py
@@ -23,6 +23,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 import typing
 import os.path
+import abc
 
 import libzfs
 
@@ -41,7 +42,7 @@ import iocage.lib.helpers
 import iocage.lib.Config.File  # noqa: F401
 
 
-class Resource:
+class Resource(metaclass=abc.ABCMeta):
     """
     iocage resource
 
@@ -123,6 +124,10 @@ class Resource:
             self.dataset_name = dataset_name
         elif dataset is not None:
             self.dataset = dataset
+
+    @abc.abstractmethod
+    def destroy(self):
+        pass
 
     @property
     def pool(self) -> libzfs.ZFSPool:
@@ -321,6 +326,9 @@ class DefaultResource(Resource):
             logger=logger
         )
 
+    def destroy(self):
+        raise NotImplementedError("destroy unimplemented for DefaultResource")
+
     def save(self) -> None:
         self.write_config(self.config.user_data)
 
@@ -348,6 +356,9 @@ class ListableResource(list, Resource):
         )
 
         self.filters = filters
+
+    def destroy(self):
+        raise NotImplementedError("destroy unimplemented for ListableResource")
 
     def __iter__(
         self

